core/src/main/scala/zquery/DataSource.scala

```diff
@@ -0,0 +1,149 @@
+    )(name: String)(f: C => Either[A, B]): DataSource.Service[R1, E1, C] =
+      new DataSource.Service[R1, E1, C] {
+        val identifier = s"${self.identifier}.eitherWith(${that.identifier})($name)"
+        def run(requests: Iterable[C]): ZIO[R1, E1, CompletedRequestMap] = {
+          val (as, bs) = requests.foldLeft((List.empty[A], List.empty[B])) {
+            case ((as, bs), c) =>
+              f(c) match {
+                case Left(a)  => (a :: as, bs)
+                case Right(b) => (as, b :: bs)
+              }
+          }
+          self.run(as).zipWithPar(that.run(bs))(_ ++ _)
+        }
+      }
+
+    override final def equals(that: Any): Boolean = that match {
+      case that: DataSource.Service[_, _, _] => this.identifier == that.identifier
+    }
+
+    override final def hashCode: Int =
+      identifier.hashCode
+
+    /**
+     * Returns a new data source with failures mapped using the specified
+     * function.
+     */
+    final def mapError[E1](name: String)(f: E => E1): DataSource.Service[R, E1, A] =
+      new DataSource.Service[R, E1, A] {
+        val identifier = s"${self.identifier}.mapError($name)"
+        def run(requests: Iterable[A]): ZIO[R, E1, CompletedRequestMap] =
+          self.run(requests).mapError(f)
+      }
+
+    /**
+     * Provides this data source with its required environment.
+     */
+    final def provide(name: String)(r: R): DataSource.Service[Any, E, A] =
+      provideSome(s"_ => $name")(_ => r)
+
+    /**
+     * Provides this data source with part of its required environment.
+     */
+    final def provideSome[R0](name: String)(f: R0 => R): DataSource.Service[R0, E, A] =
+      new DataSource.Service[R0, E, A] {
+        val identifier = s"${self.identifier}.provideSome($name)"
+        def run(requests: Iterable[A]): ZIO[R0, E, CompletedRequestMap] =
+          self.run(requests).provideSome(f)
+      }
+
+    override final def toString: String =
+      identifier
+  }
+
+  /**
+   * Constructs a data source from a pure function.
+   */
+  final def fromFunction[A, B](
+    name: String
+  )(f: A => B)(implicit ev: A <:< Request[B]): DataSource.Service[Any, Nothing, A] =
+    new DataSource.Service[Any, Nothing, A] {
+      val identifier = name
+      def run(requests: Iterable[A]): ZIO[Any, Nothing, CompletedRequestMap] =
+        ZIO.succeed(requests.foldLeft(CompletedRequestMap.empty)((map, k) => map.insert(k)(f(k))))
+    }
+
+  /**
+   * Constructs a data source from an effectual function.
+   */
+  final def fromFunctionM[R, E, A, B](
+    name: String
+  )(f: A => ZIO[R, E, B])(implicit ev: A <:< Request[B]): DataSource.Service[R, E, A] =
+    new DataSource.Service[R, E, A] {
+      val identifier = name
+      def run(requests: Iterable[A]): ZIO[R, E, CompletedRequestMap] =
+        ZIO
+          .foreachPar(requests)(k => ZIO.succeed(k).zip(f(k)))
+          .map(_.foldLeft(CompletedRequestMap.empty) { case (map, (k, v)) => map.insert(k)(v) })
+    }
+}
```

Comment:
Maybe a helper from a name and a function from requests to result map could be nice.
So this snippet:
```scala
  object UpcomingEventDataSource extends DataSource[Any, Nothing, GetUpcomingEventIdsForUser] {
    val dataSource: DataSource.Service[Any, Nothing, GetUpcomingEventIdsForUser] =
      new DataSource.Service[Any, Nothing, GetUpcomingEventIdsForUser] {
        override val identifier: String = "UpcomingEventDataSource"
        override def run(requests: Iterable[GetUpcomingEventIdsForUser]): ZIO[Any, Nothing, CompletedRequestMap] =
          UIO(println("getUpcomingEventIdsForUser")).as(requests.foldLeft(CompletedRequestMap.empty) {
            case (map, req) => map.insert(req)((1 to req.first).toList)
          })
      }
  }
```
could be written:
```scala
  val UpcomingEventDataSource = DataSource.make("UpcomingEventDataSource") { requests =>
    UIO(println("getUpcomingEventIdsForUser")).as(requests.foldLeft(CompletedRequestMap.empty) {
      case (map, req) => map.insert(req)((1 to req.first).toList)
    })
  }
```

Reply:
Yes. Added an `apply` method to let you do this.